### DISCUSSION TITLE
PointerPressed and PointerReleased events in buttons

### DISCUSSION THREAD

**Discussion by os-lmo** [AvaloniaUI/Avalonia#5450] [General]
Hi,

In the [button control](http://reference.avaloniaui.net/api/Avalonia.Controls/Button), I realized that the `PointerPressed` and `PointerReleased` events are only triggered when I press/release the mouse right button in a button. If I press the left button, only the `click` event is triggered.

My question is: why the `PointerPressed` and `PointerReleased` events are not triggered with a left mouse click in a button (keeping the mouse over the button) ?

**Comment by jp2masa** [comment]
That's because the event was already handled (`e.Handled = true`). You can subscribe to events even if they're handled like this:

```cs
button.AddHandler(Button.PointerPressedEvent, (sender, e) => { }, handledEventsToo: true);
```

**Reply by os-lmo** [reply]
Thanks. Adding the `handledEventsToo`, it worked.

Just one more question: I want to globally listening all pointer released events (just once per release) hence I am using the `AddClassHandler`:

`Window.PointerReleasedEvent.AddClassHandler((Window target, PointerReleasedEventArgs e) => { }, handledEventsToo: true);`

Anyway, these events are not being catched. I am missing anything?

**Reply by jp2masa** [reply]
That's because `routes` is `Direct|Bubble` by default (i.e. event is directly on `Window` or it was not handled by an element in the chain starting in the control and going up to the `Window`. You need to add the `Tunnel` routing strategy (like bubble, but starts from the root and goes down to the control):

```cs
button.AddHandler(
    Button.PointerPressedEvent,
    (sender, e) => { },
    routes: RoutingStrategies.Direct | RoutingStrategies.Tunnel | RoutingStrategies.Bubble,
    handledEventsToo: true);
```

**Reply by xmaxrayx** [reply]
@jp2masa  can i do it with sender object? i'm traying catch pen input so thats why i can't use click.

````cs
            sender.AddHandler(
            Button.PointerPressedEvent,
            (sender, e) => { },
            routes: RoutingStrategies.Tunnel ,
            handledEventsToo: true);

            Debug.WriteLine("pointer is pressed");
```

edit : nvr mind i forget doing x:name
